src/node_file.cc

```diff
@@ -966,32 +967,67 @@ void Access(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
-void Close(const FunctionCallbackInfo<Value>& args) {
+static void Close(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
 
-  const int argc = args.Length();
-  CHECK_GE(argc, 1);
+  CHECK_EQ(args.Length(), 2);  // fd, req
 
   int fd;
   if (!GetValidatedFd(env, args[0]).To(&fd)) {
     return;
   }
   env->RemoveUnmanagedFd(fd);
 
-  if (argc > 1) {  // close(fd, req)
-    FSReqBase* req_wrap_async = GetReqWrap(args, 1);
-    CHECK_NOT_NULL(req_wrap_async);
-    FS_ASYNC_TRACE_BEGIN0(UV_FS_CLOSE, req_wrap_async)
-    AsyncCall(env, req_wrap_async, args, "close", UTF8, AfterNoArgs,
-              uv_fs_close, fd);
-  } else {  // close(fd)
-    FSReqWrapSync req_wrap_sync("close");
-    FS_SYNC_TRACE_BEGIN(close);
-    SyncCallAndThrowOnError(env, &req_wrap_sync, uv_fs_close, fd);
-    FS_SYNC_TRACE_END(close);
+  FSReqBase* req_wrap_async = GetReqWrap(args, 1);
+  CHECK_NOT_NULL(req_wrap_async);
+  FS_ASYNC_TRACE_BEGIN0(UV_FS_CLOSE, req_wrap_async)
+  AsyncCall(
+      env, req_wrap_async, args, "close", UTF8, AfterNoArgs, uv_fs_close, fd);
+}
+
+// Separate implementations are required to provide fast API for closeSync.
+// If both close and closeSync are implemented using the same function, and
+// if a fast API implementation is added for closeSync, close(fd, req) will
+// also trigger the fast API implementation and cause an incident.
+// Ref: https://github.com/nodejs/node/issues/53902
+static void CloseSync(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_EQ(args.Length(), 1);
+
+  int fd;
+  if (!GetValidatedFd(env, args[0]).To(&fd)) {
+    return;
+  }
+  env->RemoveUnmanagedFd(fd);
+  FSReqWrapSync req_wrap_sync("close");
+  FS_SYNC_TRACE_BEGIN(close);
+  SyncCallAndThrowOnError(env, &req_wrap_sync, uv_fs_close, fd);
+  FS_SYNC_TRACE_END(close);
+}
+
+static void FastCloseSync(Local<Object> recv,
+                          const uint32_t fd,
+                          // NOLINTNEXTLINE(runtime/references) This is V8 api.
+                          v8::FastApiCallbackOptions& options) {
+  Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
+
+  uv_fs_t req;
+  FS_SYNC_TRACE_BEGIN(close);
+  int err = uv_fs_close(nullptr, &req, fd, nullptr);
+  FS_SYNC_TRACE_END(close);
+
+  if (is_uv_error(err)) {
+    options.fallback = true;
```

Comment:
This means retrying, correct? In that case, quoting https://man7.org/linux/man-pages/man2/close.2.html: (emphasis mine)

> Dealing with error returns from close()
       A careful programmer will check the return value of close(), since it is quite possible that errors on a previous [write(2)](https://man7.org/linux/man-pages/man2/write.2.html) operation are reported only on the final close() that releases the open file description.  Failing to check the return value when closing a file may lead to silent loss of data.  This can especially be observed with NFS and with disk quota.
>
> Note, however, that a failure return should be used only for diagnostic purposes (i.e., a warning to the application that there may still be I/O pending or there may have been failed I/O) or remedial purposes (e.g., writing the file once more or creating a backup).
>
> **Retrying the close() after a failure return is the wrong thing to do**, since this may cause a reused file descriptor from another thread to be closed.  This can occur because the Linux kernel always releases the file descriptor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.
>
> Many other implementations similarly always close the file descriptor (except in the case of EBADF, meaning that the file descriptor was invalid) even if they subsequently report an error on return from close().  POSIX.1 is currently silent on this point, but there are plans to mandate this behavior in the next major release of the standard.
>
> A careful programmer who wants to know about I/O errors may precede close() with a call to [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html).
>
> The EINTR error is a somewhat special case.  Regarding the EINTR error, POSIX.1-2008 says:
>
> If close() is interrupted by a signal that is to be caught, it shall return -1 with [errno](https://man7.org/linux/man-pages/man3/errno.3.html) set to EINTR and the state of fildes is unspecified.
>
> This permits the behavior that occurs on Linux and many other implementations, where, as with other errors that may be reported by close(), the file descriptor is guaranteed to be closed. However, it also permits another possibility: that the implementation returns an EINTR error and keeps the file descriptor open.  (According to its documentation, HP-UX's close() does this.)  The caller must then once more use close() to close the file descriptor, to avoid file descriptor leaks. This divergence in implementation behaviors provides a difficult hurdle for portable applications, since on many implementations, close() must not be called again after an EINTR error, and on at least one, close() must be called again.  There are plans to address this conundrum for the next major release of the POSIX.1 standard.

Reply:
Yes, `options.fallback = true` will retry on the slow path.

> Retrying the close() after a failure return is the wrong thing to do, since this may cause a reused file descriptor from another thread to be closed. This can occur because the Linux kernel always releases the file descriptor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.

Does this still apply since this is a IO blocking operation? If it does apply, does this make this pull-request invalid?

> A careful programmer who wants to know about I/O errors may precede close() with a call to [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html).

This might be a breaking change isn't it?

Reply:
> Yes, `options.fallback = true` will retry on the slow path.
 
Ok – we can't really do that.

> > Retrying the close() after a failure return is the wrong thing to do, since this may cause a reused file descriptor from another thread to be closed. This can occur because the Linux kernel always releases the file descriptor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.
> 
> Does this still apply since this is a IO blocking operation? If it does apply, does this make this pull-request invalid?

Not really – it just means we can't use a fallback path here.

> > A careful programmer who wants to know about I/O errors may precede close() with a call to [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html).
> 
> This might be a breaking change isn't it?

It's not a suggestion that I'd implement inside of our `fs.close()` calls. If a user wants this, they can call our own `fsync()` wrapper.

Reply:
when `isolate` is available on `FastApiCallbackOptions`, you'll be able to raise an exception directly from the fast function. Maybe this should wait for that?